### PR TITLE
`Exam Mode`: Add sidebar to student view to display exams

### PR DESCRIPTION
**Pull request by edkaya**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [ ] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

In the current design of Artemis, we display real and test exams side by side. In other parts of Artemis, such as lectures and exercises, we use sidebars to group elements. To maintain consistency, the sidebar is also implemented in the course exams view in order to group and display real and test exams. This implementation is therefore a part of the new design in exam mode. 

### Description

- Added the new sidebar into course exams
- Since the current implementation covers both exam-participation-cover (welcome&end page) and exam-participation components in the same route url, the problem occured in a way that both welcome page and exam itself is displayed in the sidebar content. But the intended behaviour is to display welcome-page in the sidebar content, whereas the exam itself as full-screen. To omit this problem, the approach is implemented where sidebars and breadcrum get hidden when an exam is started. Therefore, we can now see the exam in full-screen again. 
- Welcome-page, exam-summary and submission error text are displayed in sidebar-content.
- Note: Status icons in the sidebar cards for the exams in the sidebar will be implemented in a follow-up PR (See screenshots)

### Steps for Testing

Prerequisites:
- 1 Admin
- 1 Students
- 1 Real Exam
- 1 Test Exam

**(There are already three exams available under my course "Ege Dogu Kaya" on TS4 and only one test account (artemis_test_user_4) is registered for these three exams, feel free to register other accounts if needed. If you want to use these three exams to test it, you can skip first 8 steps. )**

1. Log in to Artemis with Admin account
2. Navigate to Course Management
3. Select a Course
4. Navigate to Exams
5. Create an Exam 
6. Register the student account
7. Go to Student Exams
8. Click 'Generate individual exams' and then 'Prepare exercise start'
9. Log in to Artemis with Student account
10. Select the course 
11. Navigate to Exams 
12. Verify that you have real and test exams being displayed in the sidebar with correct information
13. Verify that you can hide the sidebar with toggle feature (CTRL+Shift+B)
14. Verify that you see the welcome page of the exam in the sidebar-content 
15. Verify that you are correctly navigated to exam after clicking on 'Start'
16. Verify that you can see the exam summary in the sidebar-content when you finished the exam

#### Exam Mode Testing

Prerequisites:
- 1 Student
- 1 Exam

1. Log in to Artemis
2. Participate in the exam as a student
3. Make sure that nothing changes in the exam itself, both functionally and visually.

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| course-exams.component.ts | 91.17% | ✅                          |
| course-overview.service.ts | 92.3%  | ✅                             |

### Screenshots

- Current system
Screenshot 2024-05-26 at 02 07 04

- Intended changes according to the mockups in Figma
Screenshot 2024-05-25 at 17 11 50

- Implemented design
Screenshot 2024-05-25 at 20 00 09

## Summary by CodeRabbit

- **New Features**
  - Improved handling of exam and course components with conditional rendering and updated styling.
  - Added support for displaying exam-specific details like start time, working time, and attainable points in the sidebar.

- **Bug Fixes**
  - Corrected trailing commas in various Angular template bindings to ensure proper functionality.

- **Refactor**
  - Updated course overview and exam components with clearer, more user-friendly interfaces.
  - Enhanced logic for handling exam participation and assessment.

- **Tests**
  - Added new tests to verify functionality for exam and sidebar components handling different scenarios.
  
- **Documentation**
  - Updated translations in the German and English student dashboard JSON files for new exam-related terms.

### CODE REVIEW

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Tested on TS3, starting the exam does not seem to work any more.
image
It seems to create a new exam session for every click on the start button though, visible under the student exams overview.

**Comment by edkaya** [comment]
> Tested on TS3, starting the exam does not seem to work any more. image It seems to create a new exam session for every click on the start button though, visible under the student exams overview.

Thanks a lot for your insights! It seems I need to adapt my routing according to the new changes in the develop, because there are some new changes regarding the routing in courses module. I was going to fix this tomorrow, so exam will work again but it seems I forgot to change the label to "Work in Progress", sorry about that. I'll let you know when I am finished with it.

**Review by DominikRemo — CHANGES_REQUESTED** [review]
Please take another look at the client tests. There currently seems to be a problem with the code coverage:

> Jest: "global" coverage threshold for statements (87.34%) not met: 87.33%
Jest: "global" coverage threshold for branches (73.9%) not met: 73.85%
Jest: "global" coverage threshold for lines (87.43%) not met: 87.41%

**Review by DominikRemo — DISMISSED** [review]
Tested on ts4: Exams are displayed in the sidebar. The sidebar can be toggled on and off using CTRL+Shift+B. Starting the exam correctly takes me to the exam page.
Tested on ts5: Exam results are displayed correctly for previous exams.

Idea: In the past, open exams were easily distinguishable from closed exams. It may be useful to add an indicator in this PR or in a follow up, although this is very optional and just something I noticed.

**Comment by edkaya** [comment]
> Tested on ts4: Exams are displayed in the sidebar. The sidebar can be toggled on and off using CTRL+Shift+B. Starting the exam correctly takes me to the exam page. Tested on ts5: Exam results are displayed correctly for previous exams.
> 
> Idea: In the past, open exams were easily distinguishable from closed exams. It may be useful to add an indicator in this PR or in a follow up, although this is very optional and just something I noticed.

Thanks for your feedback. I will implement status icons in a follow up PR to indicate if the exam is open/closed or in another status. Since this process is still in discussion, I didn't implement in this PR.

**Review by muradium — APPROVED** [review]
Tested on TS4 on Safari, all interactions with sidebar and content work as expected, hotkey works 👍 

One opinion generally about the sidebar for the future - I think it would be more intuitive if sidebar was visible every time user navigates back to exams/exercises. Because it may be hard to guess where to toggle it back on or generally user may forget about its existence and can't find the list of exams/exercises 😄

**Comment by edkaya** [comment]
> Tested on TS4 on Safari, all interactions with sidebar and content work as expected, hotkey works 👍
> 
> One opinion generally about the sidebar for the future - I think it would be more intuitive if sidebar was visible every time user navigates back to exams/exercises. Because it may be hard to guess where to toggle it back on or generally user may forget about its existence and can't find the list of exams/exercises 😄

Thank you for your advice and opinion 🙂 I think it is the intended behaviour for all sidebar components such as lectures, exercises, exams and tutorial groups. Because the sidebar gets invisible if only the user manually toggles it, which means that user is aware of that feature. But since exams can be long and user may forget about this feature, I can discuss this with Stephan and Ramona about it 👍

File: src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html
Comment (rstief) [inline src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html:20]: (optional) we prefer jhiTranslate over artemisTranslate

File: src/main/webapp/app/overview/course-overview.service.ts
Comment (rstief) [inline src/main/webapp/app/overview/course-overview.service.ts:256]: Is there a reason we cannot just simplify this as follows:
```suggestion
            if (hours > 0) {
                workingTimeString += `${hours}h`;
            }
            if (minutes > 0) {
                workingTimeString += `${minutes} min`;
            }   
            if (seconds> 0) {
                workingTimeString += `${seconds} sec`;
            }
```

**Review by rstief — DISMISSED** [review]
Code LGTM overall just some small comments

File: src/main/webapp/app/overview/course-overview.service.ts
Comment (edkaya) [reply src/main/webapp/app/overview/course-overview.service.ts:256]: No, we can directly remove it. I am using artemisDate pipe in my follow up PR, so this function will removed in the follow up PR.

File: src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html
Comment (rabeatwork) [reply src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html:4]: Totally agree. @edkaya, can you please create an issue for refactoring the `sidebar-card-item.component` to make it more generic and assign it to me?

Comment (edkaya) [reply src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html:4]: Sure, I will do it now 👍

File: src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html
Comment (edkaya) [reply src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html:20]: I changed this in a follow-up PR too 😄

Comment (edkaya) [reply src/main/webapp/app/shared/sidebar/sidebar-card-item/sidebar-card-item.component.html:20]: But this file is going to be refactored by @rabeatwork anyway, so I think it would be better if I don't apply this change in this PR again

**Review by rstief — APPROVED** [review]
Code LGTM

**Review by BBesrour — APPROVED** [review]
Tested on TS4. Works as expected.
- Real and test exams being displayed in the sidebar with correct information
- Can hide sidebar with shortcut
- Can exam welcome page and navigates me to the correct exam when I click on start
- Can view summary after submission with submitted answers

Small nitpicks:
- Personally, I would prefer if the tooltip for the hide sidebar button displayed Shift instead of the arrow up, since the arrow up symbol isn't very visible on Edge.
- We could add a sliding animation when hiding the sidebar.

**Review by DominikRemo — DISMISSED** [review]
Tested again on TS4. Reapprove

**Review by rstief — DISMISSED** [review]
LGTM

**Review by BBesrour — DISMISSED** [review]
Re-approve after merge conflict

**Review by DominikRemo — DISMISSED** [review]
Reapprove

**Review by muradium — DISMISSED** [review]
Reapprove

**Review by muradium — APPROVED** [review]
Reapprove after merge conflict

**Review by DominikRemo — APPROVED** [review]
Reapprove

**Review by rstief — DISMISSED** [review]

**Review by BBesrour — DISMISSED** [review]

**Review by rabeatwork — APPROVED** [review]
maintainer approve

File: src/main/webapp/app/exam/participate/exam-participation.component.html
Comment (edkaya) [reply src/main/webapp/app/exam/participate/exam-participation.component.html:6]: Thanks! I will fix them in the follow up PR

**Comment by egekurt123** [comment]
Could you please check if the Playwright tests are running correctly? I had to make some adjustments after modifying the sidebar on the messages page, and it might be necessary to do the same here. Thank you!

**Review by DominikRemo — APPROVED** [review]
Reapprove after merge conflict

**Review by muradium — APPROVED** [review]
Reapprove after merge

**Pull request closed without merging**
